generating unique names for new tracks

szeli1 opened the pull request:
This Pull request adds 2 methods that concatenates numbers to tracks that would have the same name otherwise.

Tracks with the same name are still possible, but all new tracks and all cloned tracks will have a unique name by default. Instruments of the same kind will also have differentiating numbers added to them.

Tracks loaded from a save should not be effected.

JohannesLorenz commented:
Does this close an existing issue?

tresf commented:
Overall, this is a very nice feature.  It seems very stable and intuitive.  Here's some feedback:

Some testing:

* `4294967295`, cloned becomes `1`
* `E7`, cloned becomes `E 8` (space added)
* `a42949672951`, cloned becomes `a 1`
* `0.1` cloned becomes `0. 2` (space added)
* `1,000` cloned becomes `1,1`

Other items:

* Presets do not get renamed when added through the sidebar.  (e.g. Hit <kbd>Enter</kbd> on AudioFileProcessor -> Bass-Mania.xpf a bunch of times
* Samples do not get renamed when added through the sidebar. (e.g. Basses -> bass_acid1.ogg)
* Soundfonts do not get renamed when added through the sidebar

Some notes:
- For some reason we call the first BBEditor track "Pattern 0", but instruments do not have the zero after them.  The project may want to make this more consistent.  **Edit:** This may be saved in the default template.  Deleting "Pattern 0" and adding a new pattern, it's called "Pattern" (without the zero).

I compared this to some other software... 
- Gimp adds "#1" to identically named layers.  "Layer 1" will become "Layer 1 #1"
- GarageBand starts the vocal track at "Audio 1" so that when more are added, they're called "Audio 2", etc.
   - If the there's only one track called "Audio 2" and you create a new one, Garage Band calls it "Audio 2" still.
   - GarageBand's counter can break and get stuck.   Right now I have 5 tracks called "Audio 2".
   - GarageBand only treats vocal tracks this way, instrument tracks do not get auto numbered for some reason.

szeli1 commented:
>     4294967295, cloned becomes 1 (I would expect 4294967295 1 if integer overflow occurs)

I think this is fine. I can fix it if you think it is important.

>     E7, cloned becomes E 8 (space added -- I would expect E7 1 or possibly E8)

I think I will fix this so it will be "E8".

>     a42949672951, cloned becomes a 1.

I may look into this. I believe an ideal solution would be "a42949672951 1".

>     Presets do not get renamed when added through the sidebar. (e.g. Hit Enter on AudioFileProcessor -> Bass-Mania.xpf a bunch of times, they all will have the same name. This differs from when cloned)
>     Samples do not get renamed when added through the sidebar. (e.g. Basses -> bass_acid1.ogg)
>     Soundfonts do not get renamed when added through the sidebar

Thanks for noticing these issues

> For some reason we call the first BBEditor track "Pattern 0", but instruments do not have the zero after them. The project may want to make this more consistent. Edit: This may be saved in the default template. Deleting "Pattern 0" and adding a new pattern, it's called "Pattern" (without the zero).

So should I change the default template?

Thanks for the review tresf!

tresf commented:
My instinct is:
* yes, change the default template to remove the zero for consistency
* try to not parse the number at the end of a word unless there's a space before it.
  * this has the disadvantage of `3xosc1` becoming `3xosc1 1`
* if you still want to parse the number without spaces, maybe we change it to remember whether or not a space was there.

Not a bug, but a funny use-case I thought of yesterday but omitted from my testing:
- Call a track name "Roland 808"
- Clone track
- New track is called "Roland 809"

:D

szeli1 commented:
I fixed most of your issues: Now if the number at the end of the name is too long, it will start to count up from 1 again

>     Presets do not get renamed when added through the sidebar. (e.g. Hit Enter on AudioFileProcessor -> Bass-Mania.xpf a bunch of times, they all will have the same name. This differs from when cloned)
>     Samples do not get renamed when added through the sidebar. (e.g. Basses -> bass_acid1.ogg)
>     Soundfonts do not get renamed when added through the sidebar

I fixed this with audio file processor, but fixing this for all presets and soundfonts will be difficult, so I didn't do it.

CorruptVoidSoul commented:
I tested this too, didn't find anything that tresf didn't already mention, and I also confirm that the fixes you made so far do indeed work.

tresf commented:
> I fixed most of your issues: Now if the number at the end of the name is too long, it will start to count up from 1 again
> 
> > ```
> > Presets do not get renamed when added through the sidebar. (e.g. Hit Enter on AudioFileProcessor -> Bass-Mania.xpf a bunch of times, they all will have the same name. This differs from when cloned)
> > Samples do not get renamed when added through the sidebar. (e.g. Basses -> bass_acid1.ogg)
> > Soundfonts do not get renamed when added through the sidebar
> > ```
> 
> I fixed this with audio file processor, but fixing this for all presets and soundfonts will be difficult, so I didn't do it.

```diff
- instrumentTrack()->setName( PathUtil::cleanName( _audio_file ) );
+ instrumentTrack()->setName(instrumentTrack()->findUniqueName(PathUtil::cleanName(_audio_file)));
```

A bit of a progressive idea but what if the rename was just done for all calls to `instrumentTrack()->setName(...)`?  This would prevent a user from naming two tracks identically, but this idea isn't completely outrageous.

szeli1 commented:
> A bit of a progressive idea but what if the rename was just done for all calls to instrumentTrack()->setName(...)? This would prevent a user from naming two tracks identically, but this idea isn't completely outrageous.

Sometimes naming a thing explicitly is useful like when loading in a track from a save. I tested this idea with Instrument tracks and it broke things. It is possible but it will take more work I believe. Should I do it?

tresf commented:
> > A bit of a progressive idea but what if the rename was just done for all calls to instrumentTrack()->setName(...)? This would prevent a user from naming two tracks identically, but this idea isn't completely outrageous.
> 
> Sometimes naming a thing explicitly is useful like when loading in a track from a save. I tested this idea with Instrument tracks and it broke things. It is possible but it will take more work I believe. Should I do it?

I thought of this (many of my older tracks have identically named tracks).  I wasn't sure what type of issues it would cause.  I think that the PR -- as it stands -- is a good default improvement over the existing behavior, so without knowing how bad this breaks stuff, I'm hesitant to made this decision on your behalf.  Would you prefer to try this and see how breaking it is?  I don't want to send good work down a rabbit hole on a whim or a hunch. 😄

szeli1 commented:
> I'm hesitant to make this decision on your behalf. Would you prefer to try this and see how breaking it is

I implemented it, please review and test it, especially for the plugins (vestige, zyn, montero, gigPlayer, OpulenZ, Patman, Sf2Player)

tresf commented:
> > I'm hesitant to make this decision on your behalf. Would you prefer to try this and see how breaking it is
> 
> I implemented it, please review and test it, especially for the plugins (vestige, zyn, montero, gigPlayer, OpulenZ, Patman, Sf2Player)

This is awesome.  I haven't tested it yet, but we can likely simplify the amount of lines (~18) changed by having `setName` default to `true` and only need `false` when the default behavior is being overriden.  I'm also curious the reasoning behind some of the code that doesn't use this technique.  For example, what's the concern with things like MIDI import automatically numbering the tracks?  Assuming this is for accuracy purposes with the MIDI track, I would at least expect the time signature and tempo items that we create to auto-number.  Are there downsides to this?

szeli1 commented:
>  For example, what's the concern with things like MIDI import automatically numbering the tracks?

I didn't knew what it did and how to make it execute. There was a recursion elsewhere (maybe SampleTrackWindow) that caused a crash.

>  I haven't tested it yet, but we can likely simplify the amount of lines (~18) changed by having setName default to true and only need false when the default behavior is being overriden

I think this way is better, because a developer will have to know and make a decision about using unique names when using `setName`. This way bugs can be avoided like the recursion I mentioned.
